pkg/gatewayserver/io/ws/ws.go

```diff
@@ -254,8 +254,13 @@ func (s *srv) handleTraffic(c echo.Context) (err error) {
 		defer ticker.Stop()
 	}
 
-	go func() {
+	go func() (err error) {
 		defer ws.Close()
+		defer func() {
+			if err != nil {
+				conn.Disconnect(err)
+			}
+		}()
```

Comment:
the `ws.Close()` causes the uplink flow to disconnect which returns to the defer in `L217`. I think that this explicit disconnection is not needed. 
If you mean to sync the downstream and upstream loops, i.e., exit the upstream loop on downstream error explicitly, then the upstream loop needs to listen on the connection context right?

Reply:
The intent here is to have the error of the downstream messages propagated to the error of the connection itself. If we do only `ws.Close()`, then the error that cancels the connection context will be that the connection has been closed, not that an error was encountered during the downstream flow. Does this make sense ?

Reply:
Ok yeah that sounds good.